src/destreamer.ts

```diff
@@ -106,7 +106,7 @@ function extractVideoGuid(videoUrls: string[]): string[] {
     for (const url of videoUrls) {
         try {
             guid = url.split('/').pop();
-
+            guid = guid != undefined ? guid.split('?')[0] : guid; // Account for URLs of the form: https://web.microsoftstream.com/video/<videoId>?channelId=<channelId>
```

Comment:
ok, the original code is probably wrong, but this could be a nested bug.

first of all, i don't really know why this should throw, since pop and split never throw, they just give you the wrong input if they fail ( @snobu @lukaarma )
never hit this, can undefined throw?

about your code:

what if there s no query string? could this throw because [0] is invalid?
could this break a correct guid somehow?

Reply:
We have a URL module in Node, why not use that instead of regexing our way into oblivion? :) https://nodejs.org/api/url.html

Reply:
const myURL = new URL('https://example.org/abc/xyz?123');
console.log(myURL.pathname);
// Prints /abc/xyz

Reply:
yes, this is the way to get the pathname, but we still need a regex to get the guid from pathname, i think, and i'm suppoing this will never throw as is, so we must also check the guid is valid and then throw

URL will be a proper fix, part 1, regex and guid check is part 2 :)

Reply:
I agree using a URL module is probably the best way to go. 

When I tried just `guid.split('>')`, TypeScirpt wouldn't compile giving me something along the lines of `Object is possibly 'null' or 'undefined'`

Reply:
> what if there s no query string? could this throw because [0] is invalid?
> could this break a correct guid somehow?

I tried a correct GUID it works fine. This is because `"x".split("?")` returns `["x"]` whereas `"x?y".split("?")` returns `["x", "y"]`. The zero-th element is always the one we want.

Reply:
@kylon, but since we have a clean pathname without query parameters, isn't it safe to assume whatever comes after the last slash is our GUID?

Reply:
yes, if we suppose this, we should just `if (undefined) throw`, that should give us the correct error.
we can handle more urls if/when someone will open an issue

Reply:
> 
> 
> We have a URL module in Node, why not use that instead of regexing our way into oblivion? :) https://nodejs.org/api/url.html

Here I made [an example using the URL module](https://repl.it/repls/QueasyYearlyCylinder). It needs to be translated to TypeScript.

```javascript
const url = 'https://web.microsoftstream.com/video/VIDEO-ID?channelId=CHANNEL-ID';
const urlObj = new URL(url);
console.log(urlObj);
const videoId = urlObj.pathname.split('/').pop();
console.log(videoId);
```

Reply:
http://jsfiddle.net/Ng4UA/26/ We can throw that in if we absolutely need to make sure it's a GUID but i strongly believe we should just take the last URL element as our GUID for now.

Reply:
no, ok, we will add it if we find someone that needs this.

```
const urlObj = new URL(url); // add try catch, this can throw and we like cute error messages :D
console.log(urlObj);
const videoId = urlObj.pathname.split('/').pop();
console.log(videoId);
```

what s the typescript error with this?

Reply:
@snobu 
mmh, do we really need the query string?

Reply:
> ```
> const urlObj = new URL(url); // add try catch, this can throw and we like cute error messages 
> ```
> what s the typescript error with this?
> @constructor-s

I didn't add that comment and I don't know what are the possible errors. I was just saying I wrote it my snippet in JavaScript and maybe some changes needs to be made to translate to TypeScript. Sorry if it was confusing.

Reply:
We don't care about any query strings but they may be passed in the CLI params or the text file (-f) by the user. The first thing we should do is sanitize that input, which we already do .. or we did but had a regression.

Reply:
> 
> 
> We don't care about any query strings but they may be passed in the CLI params or the text file (-f) by the user. The first thing we should do is sanitize that input, which we already do .. or we did but had a regression.

Exactly and this is what made my CLI command fail. Not because the query string was necessary. I can manually sanitize the input argument/txt of course, but that is one more thing to specify in README.

Reply:
@constructor-s 
yeah, the comment is mine : P
i though you were updating your PR

we do sanitize the urls, but we don't remove the query string because it may be needed to open the page in puppeteer, but after that, at this point, we can just remove all the query strings in download function as a prepare step

Reply:
`because it may be needed to open the page in puppeteer` - we should not care about that, the MS Stream site will redirect the user to whatever OAuth/ADFS provider they need to log on to, it doesn't care about query params.

Reply:
> 
> 
> @constructor-s
> yeah, the comment is mine : P
> i though you were updating your PR
> 

Ah I see. The block was already in a try-catch actually. 

Updated pull request.

Reply:
ok, i did nothing with the queries because you said something that i don't remember right now, if we don't care, i will just remove them in sanitize function

Reply:
Could be a misunderstanding, we were talking about a query parameter that we get redirected to from the identity provider, and if it's there then we can't evaluate sessionInfo yet. However that has nothing to do with out input URLs which should be clean of any parameters.

Reply:
ok, good
i ll remove the queries